src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp

```diff
@@ -0,0 +1,188 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+#include "dfcc_root_object.h"
+
+#include <util/byte_operators.h>
+#include <util/cprover_prefix.h>
+#include <util/expr.h>
+#include <util/expr_util.h>
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/pointer_expr.h>
+#include <util/std_code.h>
+
+#include <goto-programs/pointer_arithmetic.h>
+
+#include <langapi/language_util.h>
+
+/// Recursively computes a set of root object expressions for \p expr.
+static void root_objects_rec(
+  const exprt &expr,
+  const namespacet &ns,
+  messaget log,
+  std::unordered_set<exprt, irep_hash> &dest)
+{
+  if(expr.id() == ID_symbol)
+  {
+    dest.insert(expr);
+  }
+  else if(expr.id() == ID_index)
+  {
+    root_objects_rec(to_index_expr(expr).array(), ns, log, dest);
+  }
+  else if(expr.id() == ID_member)
+  {
+    const typet &type = to_member_expr(expr).struct_op().type();
+    if(
+      type.id() == ID_struct || type.id() == ID_struct_tag ||
+      type.id() == ID_union || type.id() == ID_union_tag)
+    {
+      root_objects_rec(to_member_expr(expr).compound(), ns, log, dest);
+    }
+    else
+    {
+      throw unsupported_operation_exceptiont(
+        "unexpected assignment to member of '" + type.id_string() + "'");
+    }
+  }
+  else if(expr.id() == ID_if)
+  {
+    root_objects_rec(to_if_expr(expr).true_case(), ns, log, dest);
+    root_objects_rec(to_if_expr(expr).false_case(), ns, log, dest);
+  }
+  else if(expr.id() == ID_typecast)
+  {
+    root_objects_rec(to_typecast_expr(expr).op(), ns, log, dest);
+  }
+  else if(
+    expr.id() == ID_byte_extract_little_endian ||
+    expr.id() == ID_byte_extract_big_endian)
+  {
+    root_objects_rec(to_byte_extract_expr(expr).op(), ns, log, dest);
+  }
+  else if(expr.id() == ID_complex_real)
+  {
+    root_objects_rec(to_complex_real_expr(expr).op(), ns, log, dest);
+  }
+  else if(expr.id() == ID_complex_imag)
+  {
+    root_objects_rec(to_complex_imag_expr(expr).op(), ns, log, dest);
+  }
+  else if(
+    const auto &deref =
+      expr_try_dynamic_cast<dereference_exprt>(skip_typecast(expr)))
```

Comment:
The use of `skip_typecast` seems unnecessary here? You are recursing for a typecast in an earlier `if` case, so you would never get to see a typecast here.

Reply:
indeed